### ISSUE TITLE
Support for non-python implementations

### ISSUE THREAD

**Issue by alexklibisz** [harsha-simhadri/big-ann-benchmarks#20]
Hi, I have a question about using this framework w/ a non-python ANN implementation. 

It looks like this is mostly a fork of ann-benchmarks. So the only option for using outside of Python is to hack together a client/server setup, as has been done for a few algos in ann-benchmarks. This obviously handicaps and complicates non-python implementations, as it introduces costs of context-switching, serialization, and data transfer among processes. 

I asked and was told early on by project organizers that the big-ann challenge would support non-python implementations:

![image](https://user-images.githubusercontent.com/8015228/129908445-a5515724-0e5c-4b63-b81a-27e63c50b3a9.png)

So I'm wondering if there has been progress here, or any idea of how it might work?

It seems like it wouldn't be terribly difficult to refactor the code so that the containers executed by runner.py can have any entrypoint, e.g., a program in another language. The interface between runner and algorithm would then simply be some standard file format for inputs and nearest neighbor results. If that sounds like a good idea I can try to implement it. Otherwise maybe we can use this ticket for discussing alternatives.

Thanks
-Alex

**Comment by gosha1128** [comment]
Alex,

Thanks for bringing up this issue again here.

One of the primary advantages to the evaluation framework is that it
standardizes and normalizes the process by which all participants are
measured.  As an evaluator for the challenge, it certainly makes my job
easier compared to a situation where each participant submitted a different
code-base.

That said, I understand your concern about the "overhead" of merging
different language run-times.

Do you have an estimate of the percentage of this "overhead" compared to
the latency of a large batch query search set?  For example, the query set
for Deep-1B has 10,000 vectors.

If we could figure out a way to account for this "overhead," would that
alleviate your concerns about this?

**Comment by alexklibisz** [comment]
Hi George, thanks for the quick response.

> One of the primary advantages to the evaluation framework is that it
standardizes and normalizes the process by which all participants are
measured.  As an evaluator for the challenge, it certainly makes my job
easier compared to a situation where each participant submitted a different
code-base.

I understand the concern in adding complexity.

To clarify a bit, I'm proposing two changes:
- instead of implementing a specific Python class/interface, algorithms implement a specific CLI, which could even be identical to the Python CLI
- instead of keeping inputs/outputs (vectors/neighbors) in memory in the Python runtime, the CLI reads/writes the vectors/neighbors using the file format already used for datasets in the big-ann challenge.

The CLI still runs in a docker container, so we can just as precisely measure the container running time, the CPU/mem used, etc. 

The Dockerfile installs an artifact containing the CLI, fundamentally no different than the way the Dockerfiles currently install a Python library. If anything, this actually removes a fair amount of code and complexity from the benchmarks repo. 

> Do you have an estimate of the percentage of this "overhead" compared to
the latency of a large batch query search set?

I'll re-run the ann-benchmarks this weekend and get back to you with a number.

> If we could figure out a way to account for this "overhead," would that
alleviate your concerns about this?

Maybe, but it's still a more complex implementation. I'll have a better answer once I re-run those benchmarks.

**Comment by maumueller** [comment]
Thanks for the suggestion. I don't know why you would need some client/server setup instead of just using subprocess calls to support any cli language?

I am happy with providing the filenames of the groundtruth files to the search operations such that you can write a very simple wrapper that just uses subprocess calls. I think this is what Matthijs' comment referred to.  You could write the results to disk (or keep them in memory and pipe), and then transform them in the `get_results` methods to the format expected by the framework. If the overhead of doing that is too large, we could split up the search phase into "prepare" and "search" so that we minimize the overhead. 

I am very concerned about putting all the effort of file management/file preparation/result file writing onto the client side. Morevoer, once we go done this path, all implementations will become black boxes that will be difficult to check (and re-use for others).

**Comment by alexklibisz** [comment]
I'm confused about how subprocess calls would work. I don't want to start a new instance of some program for every vector. That would be even more wasteful than a client/server setup. I want to boot up a program that reads vectors from a file in the standard competition format and writes results to a file in the standard competition setup.

**Comment by alexklibisz** [comment]
I made a demo PR over in https://github.com/alexklibisz/ann-benchmarks/pull/1 to demonstrate the overhead of a client/server model.

Copying from the PR:

To summarize: the total overhead for a model w/ recall@100 = 0.91 on the Fashion Mnist dataset is 40%.

Unless I've made some error (please check my math and details in the PR), I think this result shows there is clear value in allowing language-specific benchmark implementations.

**Comment by maumueller** [comment]
Sorry for the late response @alexklibisz, I've been on vacation. 

> The dockerfile for the non-python algo installs an executable and sets it as the docker entrypoint. The executable should implement the same CLI as the one in the run_from_cmdline method in runner.py.

> The runner program starts the container with the standard CLI args.

> The executable basically re-implements the logic from the run method from runner.py and produces an hdf5 file with the same properties.

Sure, that would work after changing some of the command line arguments, I'm still not sure that it's worth replicating all the logic. 

What I meant above is that the python runner and your implementation just communicate via stdin/stdout; from the python runner's side (`>>>` stdout, `<<<` stdin) this could look like this:

```
>>> BUILD_INDEX <datasetname> <path_to_dataset> # build index, if necessary
<<< DONE BUILD_INDEX
>>> LOAD_INDEX  <datasetname> <path_to_dataset> # load index
<<< DONE LOAD_INDEX
>>> PREPARE <path_to_queries> # load query set from file [maybe unfair advantage over the actual wrappers because they usually copy code as well]
<<< DONE PREPARE
5 times do:
>>> QUERY 
<<< DONE QUERY # measure running time
Take best running time, or median,
>>> LIST RESULTS 
<<< 1 neighbor_1, neighbor_2, ....
<<< 2 neighbor_1, neighbor_2, ...
```
Your process is run exactly once. 

I could implement a driver for this next week.

**Comment by alexklibisz** [comment]
Hmm. Thanks for explaining it. I don't think I see how stdin/stdout is better than having a CLI that mimics the python runner and runs end-to-end. Managing a stdin/stdout event loop makes the program more complicated compared to just parsing CLI params and running. The container overall is more complicated because it requires coordinating two processes. And it could still cheat. For example, it could go ahead and start running queries before it emits "DONE PREPARE". It's a tricky problem, but I'm still thinking a CLI-based abstraction is simpler to implement and requires the same trust (via code review) model.

So far I've been able to make some progress on a branch that just mimics the CLI and runs all the way through in ann-benchmarks (i.e., Erik's project). The trickiest part was honestly just getting HDF5 to work from Java. Once that's stable I'll try the same pattern on the big-ann-benchmarks, which luckily have a much easier file format to work with from the JVM.

**Comment by maumueller** [comment]
I think that it would be much easier to write _and_ verify wrappers with the stdin/stdout approach. (I expect the wrapper will be smaller than your efforts to write the hdf5 file correctly, without replicating any of the other necessary logic.) I'll happily review your code and write another wrapper that follows my suggestion once you are happy with yours. 

Also `ann-benchmarks` has support for what I outlined above, but it's a bit more difficult to implement because of the file reading. It's documented here <https://github.com/erikbern/ann-benchmarks/tree/master/protocol> and used, for example, here <https://github.com/maumueller/mih>.

**Comment by alexklibisz** [comment]
You're right that it's quite difficult to replicate the runner and get the details correct. It's not really a lot of code, but there's some subtlety to how the metrics are computed. I started seeing good results today after maybe ~30 hours total getting this to work the past few weeks. It does make a big difference; even without much optimization on the JVM side, so far it's ~1.5-2x faster than the client/server model. 

So I'm starting to agree your idea of passing some minimal commands back and forth and reading the data from a file will be overall simpler. Or maybe the grass just looks greener now that I've tried to re-implement the runner.

The only part that seems a bit over-complicated is the last step, where it prints the results to stdout. Is there a reason the non-python process couldn't just write the results to a file?

So instead of:

```
>>> LIST RESULTS 
<<< 1 neighbor_1, neighbor_2, ....
<<< 2 neighbor_1, neighbor_2, ...
```

We would do:

```
>>> WRITE_RESULTS <path_to_results>
```

**Comment by maumueller** [comment]
Let me try to make a demo wrapper next Monday or Tuesday and then we can see how well it works. 

 > So instead of:
> ```
> >>> LIST RESULTS 
> <<< 1 neighbor_1, neighbor_2, ....
> <<< 2 neighbor_1, neighbor_2, ...
> ```
> We would do:
>
> `>>> WRITE_RESULTS <path_to_results>`

I'm not sure what you mean by results here, because the whole hdf5 file will be constructed from the runner. This is really just meant to report the nearest neighbors retrieved for the queries, nothing else. I am not sure I see much of a different on your side if you somehow have to write them to a file, or just iterate through them and print them on stdout. This is not meant to be part of the timings.

**Comment by alexklibisz** [comment]
I'm also working on a wrapper that uses basically this exact command/response format via JSON-over-HTTP instead of stdin/stdout. So it's not passing data around, but just commands with pointers for where to find the data on the local filesystem. It's also not trivial, but probably the least bad format I've tried so far. I started on stdin/stdout but found I was basically re-implementing an http client/server model. I've set the ann-benchmarks implementation aside for now. Hopefully will be ready for a PR early next week.
